app/models/theme.rb

```diff
@@ -828,23 +836,44 @@ def scss_variables
     contents
   end
 
-  def convert_settings
-    settings.each do |setting|
-      setting_row = ThemeSetting.where(theme_id: self.id, name: setting.name.to_s).first
+  def migrate_settings
+    runner = ThemeSettingsMigrationsRunner.new(self)
+    results = runner.run
 
-      if setting_row && setting_row.data_type != setting.type
-        if (
-             setting_row.data_type == ThemeSetting.types[:list] &&
-               setting.type == ThemeSetting.types[:string] && setting.json_schema.present?
-           )
-          convert_list_to_json_schema(setting_row, setting)
-        else
-          Rails.logger.warn(
-            "Theme setting type has changed but cannot be converted. \n\n #{setting.inspect}",
-          )
-        end
+    return if results.blank?
+
+    old_settings = self.theme_settings.pluck(:name)
+    self.theme_settings.destroy_all
+
+    final_result = results.last
+    final_result[:settings_after].each do |key, val|
+      self.update_setting(key.to_sym, val)
+    rescue Discourse::NotFound
+      if old_settings.include?(key)
+        final_result[:settings_after].delete(key)
+      else
+        raise Theme::SettingsMigrationError.new(
+                I18n.t(
+                  "themes.import_error.migrations.unknown_setting_returned_by_migration",
+                  name: final_result[:original_name],
+                  setting_name: key,
+                ),
+              )
       end
     end
+
+    results.each do |res|
+      record =
+        ThemeSettingsMigration.new(
+          theme_id: self.id,
+          version: res[:version],
+          name: res[:name],
+          theme_field_id: res[:theme_field_id],
+        )
+      record.calculate_diff(res[:settings_before], res[:settings_after])
+      record.save!
+    end
+    self.save!
```

Comment:
Should all of these be wrapped in a transaction? Otherwise, there is a chance that the value of settings are changed by the user while migration is running. I'm not sure if we need a lock around updating settings but I definitely think this needs to happen in a transaction.

Reply:
So this method is called from one place only and that place already has a transaction open when it calls this method. Nested transactions worry me because in some cases a child transaction may rollback without the parent rolling back and that may lead to bad state. Is it reasonable to expect a method to be called from within a transaction only?

Reply:
> Is it reasonable to expect a method to be called from within a transaction only?

In this case I'm wondering if we should just combine both methods cause this method itself has to be called within a transaction.

Reply:
The calling methods is already quite big (over 150 lines) so adding this to it will make it ungodly... maybe a way to go about this is to have a `start_transaction` option here that defaults to true, and then in the place that currently calls this method can pass false for that option?

Reply:
Yup sounds good :+1:

Reply:
>  Nested transactions worry me because in some cases a child transaction may rollback without the parent rolling back and that may lead to bad state.

Actually can you confirm this statement? I think if a child transaction rolls back, it is guaranteed that the parent transaction will not be committed as well.

Reply:
> Actually can you confirm this statement? I think if a child transaction rolls back, it is guaranteed that the parent transaction will not be committed as well.

No, there's not 100% guarantee. For example, if an `ActiveRecord::Rollback` error is raised within a child transaction, only the child transaction will rollback because the error is absorbed and never propagated to the parent transaction. Other errors do propagate and cause the parent transaction to rollback unless something rescues the error.

Reply:
> The calling methods is already quite big (over 150 lines) so adding this to it will make it ungodly... maybe a way to go about this is to have a `start_transaction` option here that defaults to true, and then in the place that currently calls this method we can pass false for that option?

I've made this change.